experimental/Schemes/WeilDivisor.jl

```diff
@@ -0,0 +1,306 @@
+      CoefficientRingElemType<:AbstractAlgebra.RingElem
+    }
+
+A Weil divisor on a covered scheme ``X`` of type 
+`CoveredSchemeType` as a formal linear combination of 
+ideal sheaves of type `IdealSheafType` with coefficients 
+in a ring ``R`` of type `CoefficientRingType`.
+"""
+@attributes mutable struct WeilDivisor{
+    CoveredSchemeType<:AbsCoveredScheme, 
+    CoefficientRingType<:AbstractAlgebra.Ring, 
+    CoefficientRingElemType<:AbstractAlgebra.RingElem
+   }
+  X::CoveredSchemeType # the parent
+  R::CoefficientRingType # the ring of coefficients
+  coefficients::Dict{IdealSheaf, CoefficientRingElemType} # the formal linear combination
+
+  function WeilDivisor(
+      X::AbsCoveredScheme,
+      R::CoefficientRingType, 
+      coefficients::Dict{<:IdealSheaf, CoefficientRingElemType}
+    ) where {CoefficientRingType, CoefficientRingElemType}
+    # TODO: Do we want to require that the different effective divisors 
+    # have the same underlying covering? Probably not.
+    for D in keys(coefficients)
+      scheme(D) == X || error("component of divisor does not lay in the given scheme")
+      parent(coefficients[D]) == R || error("coefficients do not lay in the given ring")
+    end
+    return new{typeof(X), CoefficientRingType, CoefficientRingElemType}(X, R, coefficients)
+  end
+end
+
+### type getters 
+scheme_type(D::WeilDivisor{S, U, V}) where{S, T, U, V} = S
+scheme_type(::Type{WeilDivisor{S, U, V}}) where{S, T, U, V} = S
+coefficient_ring_type(D::WeilDivisor{S, U, V}) where{S, T, U, V} = U
+coefficient_ring_type(::Type{WeilDivisor{S, U, V}}) where{S, T, U, V} = U
+coefficient_type(D::WeilDivisor{S, U, V}) where{S, T, U, V} = V
+coefficient_type(::Type{WeilDivisor{S, U, V}}) where{S, T, U, V} = V
+
+### getter methods
+scheme(D::WeilDivisor) = D.X
+getindex(D::WeilDivisor, I::IdealSheaf) = (D.coefficients)[I]
+components(D::WeilDivisor) = [ Z for Z in keys(D.coefficients)]
+coefficient_dict(D::WeilDivisor) = D.coefficients
+coefficient_ring(D::WeilDivisor) = D.R
+
+set_name!(X::WeilDivisor, name::String) = set_attribute!(X, :name, name)
+name_of(X::WeilDivisor) = get_attribute(X, :name)::String
+has_name(X::WeilDivisor) = has_attribute(X, :name)
+
+function setindex!(D::WeilDivisor, c::RingElem, I::IdealSheaf)
+  parent(c) == coefficient_ring(D) || error("coefficient does not belong to the correct ring")
+  coefficient_dict(D)[I] = c
+end
+
+function WeilDivisor(X::CoveredScheme, R::Ring)
+  D = Dict{IdealSheaf, elem_type(R)}()
+  return WeilDivisor(X, R, D)
+end
+
+function WeilDivisor(I::IdealSheaf, R::Ring)
+  D = WeilDivisor(scheme(I), R)
+  D[I] = one(R)
+  return D
+end
+
+function WeilDivisor(I::IdealSheaf)
+  D = WeilDivisor(scheme(I), ZZ)
+  D[I] = one(ZZ)
+  return D
+end
+
+### copy constructor
+# Only copies the coefficient dictionary; everything else has parent character.
+function WeilDivisor(D::WeilDivisor) 
+  return WeilDivisor(scheme(D), coefficient_ring(D), copy(coefficient_dict(D)))
+end
```

Comment:
For most types `T` this is a noop and `copy` is used for copies like that.

Reply:
Good to know. That is also relevant for another issue we discovered today.

Reply:
Or maybe rather `deepcopy`.